README.md
![Core Theory logo](assets/static/images/CT_Logo_Text.png)

# Pwned by Core Theory

![Master](https://github.com/coretheory/pwned_coretheory/workflows/Master/badge.svg?branch=master)
![Staging](https://github.com/coretheory/pwned_coretheory/workflows/Staging/badge.svg)
[![Hex.pm](https://img.shields.io/hexpm/v/pwned_coretheory.svg)](https://hex.pm/packages/pwned_coretheory)
[![Docs](https://img.shields.io/badge/hex-docs-green.svg)](https://hexdocs.pm/pwned_coretheory/readme.html)
[![Downloads](https://img.shields.io/hexpm/dt/pwned_coretheory.svg)](https://hex.pm/packages/pwned_coretheory)


A simple application to check if an email or password has been pwned using the HaveIbeenPwned? API v3.

### Emails

This library currently implements simple email checking against data breaches with the HaveIBeenPwned? API v3. It requires a [purchased api-key](https://haveibeenpwned.com/API/Key) in order to work.

The `Pwned.check_email/1` function returns the total number of times an email address has appeared in known data breaches, or an `"email not pwned"` message.

### Passwords

This library uses [HaveIBeenPwned?](https://haveibeenpwned.com) to verify if a password has appeared in a data breach. 

In order to protect the value of the source password being searched, the value is not sent through the network. Instead it uses a [k-Anonymity](https://en.wikipedia.org/wiki/K-anonymity) model that allows a [password to be searched for by partial hash](https://haveibeenpwned.com/API/v2#SearchingPwnedPasswordsByRange). This allows the first 5 characters of a SHA-1 password hash to be passed to the API. Then, it searches the results of the response for the presence of the source hash. If the source hash is not found, then the password does not exist in the data set.

Additionally, we implement padding to further protect the privacy of the password source hash in accordance with [password padding](https://haveibeenpwned.com/API/v3#PwnedPasswordsPadding) in API v3.

## Table of Contents

-   [Dependencies](#dependencies)
-   [Install](#install)
-   [Usage](#usage)
-   [Changelog](#changelog)
-   [Contributing](#contributing)
-   [Further Reading](#further-reading)
-   [License](#license)
-   [Notice](#notice)
-   [Special Thanks](#special-thanks)

## Dependencies

This package requires [httpoison v1.8](https://hex.pm/packages/httpoison). If you have v1.7 in your `mix.lock` file, then you will need to update it to `1.8` to successfully run `mix deps.get`.

It also requires Elixir 1.11 to work. If you need functionality for earlier versions of Elixir, then we'd be happy to receive a PR.

## Install

**Recommended**

This package can be installed by adding `:pwned_coretheory` to your list of dependencies in `mix.exs`:

```elixir
defp deps do
  [
    ...
    {:pwned_coretheory, "~> 1.5"},
  ]
end
```

Then, run `mix deps.get`. Additionally, run `mix deps.update pwned_coretheory` occasionally to ensure you have the latest release.

**Running on master**

If you would like to run on the master branch, then update your dependencies as such:

```elixir
defp deps do
  [
    ...
    {:pwned_coretheory, github: "coretheory/pwned_coretheory"},
  ]
end
```

## Usage

Usage is incredibly simple and straightforward. You can check if an
email or password has been pwned with calls to their respective
functions.

In the case of checking for an email, you will need to have purchased
a [hibp-api-key](https://haveibeenpwned.com/API/Key).

You can use this library for password checking without the need for an
API key. However, if this is the case, then keep in mind that not all
tests will pass. If you do not need email checking, then we encourage
you to use the [pwned](https://github.com/thiamsantos/pwned) library
by [Thiago Santos](https://github.com/thiamsantos).

### Check for pwned passwords

To check whether a password has been pwned you can make a simple call to the `Pwned.check_password/1` function:

```elixir
iex> Pwned.check_password("P@ssw0rd")
      {:ok, 47205}

iex> Pwned.check_password("Z76okiy2X1m5PFud8iPUQGqusShCJhg")
      {:ok, false}
```

When implementing in an application, we can use a straightforward `case` statement like this:

```elixir
case Pwned.check_password("somepassword") do
  {:ok, false} ->
    IO.puts("Good news — no pwnage found!")

  {:ok, count} ->
    IO.puts("Oh, no! This password appeared #{count} times in data breaches.")

  :error ->
    IO.puts("Something went wrong.")
end
```

### Check for pwned emails

First, let's make sure our `hibp-api-key` is ready to go.

**Purchase your hibp-api-key and add it to an environment file**

You will first need to purchase a `hibp-api-key` from [haveibeenpwned?](https://haveibeenpwned.com/API/Key).

Then, create a `.env` file at the root of your project (e.g. beside your `.gitignore`. Be sure to update
your `.gitignore` file to ignore environment files: `*.env`. 

Once you are certain that you will not be pushing your environment files up to a source control repository,
add your purchased `hibp-api-key` to your `.env` file: `export HIBP_API_KEY=your_hibp_api_key`. Next, you'll
want to run `source .env` from your terminal.

For production, you'll want to have your `hibp-api-key` safely stored in your production host's environment
variables configuration with the key: `HIBP_API_KEY`.

Lastly, you can easily configure the `:user_agent` for the HIBP API, like so:

```elixir
# In your config.exs.
config :pwned_coretheory,
  user_agent: "YourApp Pwned Client"
```

_We highly recommend you set the configuration as it is good practice and informs the HaveIBeenPwned? service
that it is your application accessing the data and not a spammer or malicious account._

**Checking emails**

To check whether an email has been pwned you can make a simple call to the `Pwned.check_email/1` function:

```elixir
iex> Pwned.check_email("test123@example.com")
    {:pwned_email, 4893554722}

iex> Pwned.check_email("Z76okiy2X1m5PFud8iPUQGqusShCJhg@example.com")
    {:safe_email, "email not pwned"}
```

When implementing in an application, we can use a straightforward `case` statement like this:

```elixir
case Pwned.check_email("test123@exmaple.com") do
  {:safe_email, message} ->
    IO.puts(message)

  {:pwned_email, pwned_count} ->
    IO.puts("Ohh, sorry! This email has appeared #{pwned_count} times in data breaches.")

  {:error, message} ->
    IO.puts("An error occurred: " <> message)
  
  :error ->
    IO.puts("Something went wrong.")
end
```

## Changelog

See the [changelog file](CHANGELOG.md).

## Contributing

See the [contributing file](CONTRIBUTING.md).

## Further Reading

See the [further reading file](FURTHER_READING.md).

## License

[Apache License, Version 2.0](LICENSE.md) © 2021 [Core Theory, Inc.](https://github.com/coretheory)

## Notice

This is a modified version of the [pwned](https://github.com/thiamsantos/pwned) package © [@thiamsantos](https://github.com/thiamsantos).

## Special thanks

This extension was built from the simple and wonderful package, [pwned](https://github.com/thiamsantos/pwned), by [Thiago Santos](https://github.com/thiamsantos). ♥


CHANGELOG.md
# Changelog

This is the changelog for Pwned by Core Theory.

## 1.5.3 (2022-10-17)

### Fixes

- Add proper Have I Been Pwned? test emails per their API documentation.
- Change module attribute for `@hibp_api_key` to variable.
- Update dependencies.

## 1.5.2 (2021-01-30)

### Enhancements

- Add clearer test coverage for both `master` and `staging` branches to `README.md`.
- Refactor GitHub Actions workflow.
- Improve function docs clarity for `EmailFlattener.flatten/1` and `EmailReducer.reduce_email_list/1`.
- Add a security policy.

## 1.5.1 (2021-01-27)

### Enhancements

- Format changelog.
- Add further reading section with notes on rate limiting and additional resources.

## 1.5.0 (2021-01-27)

### Enhancements

- Replace unsafe runtime environment call with compile time call in `api_client.ex`.
- Refactor `api_client.ex` for clarity, performance, and maintainability.

## 1.2.4 (2021-01-26)

### Enhancements

- Add `hibp-api-key` to CI to enable checking email and password tests.

## 1.2.3 (2021-01-26)

### Enhancements

- Add default for application environment `:user_agent` and ability to override in 3rd party application config files.
- Add usage explanation for setting the `:user_agent` in 3rd party applications.
- Add `@moduledoc` to `Pwned.APIClient`.

## 1.2.2 (2021-01-26)

### Enhancements

- Add `CHANGELOG.md` file.

## 1.2.1 (2021-01-26)

### Bug fixes

- Update application name to match `:pwned_coretheory`. Previous application name mismatch was causing `Pwned.check_email(email)` to return `{:error, "unauthorized api key"}` with valid key.

test/test_helper.exs
ExUnit.start()
Application.put_env(:pwned_coretheory, :range_client, Pwned.Range.Sandbox)


test/pwned/email_test.exs
defmodule Pwned.EmailTest do
  @moduledoc """
  Tests that the check_email/1 function returns the correct
  tuple using the HaveIBeenPwned? API v3.

  This requires a purchased hibp-api-key. You can purchase one
  here: https://haveibeenpwned.com/API/Key.
  """
  use ExUnit.Case, async: true
  alias Pwned

  @test_pwned_email "multiple-breaches@hibp-integration-tests.com"
  @test_safe_email "not-active-breach@hibp-integration-tests.com"

  describe "verify check_email/1" do
    test "check_email returns the email pwned count when email is pwned" do
      assert {:pwned_email, pwned_count} = Pwned.check_email(@test_pwned_email)
      assert pwned_count > 0
    end

    test "check_email pwned count is an integer" do
      {:pwned_email, pwned_count} = Pwned.check_email(@test_pwned_email)

      assert is_integer(pwned_count)
    end

    test "check_email returns the :safe_email message when email is not pwned" do
      # This may fail if this test email appears in a future databreach.
      assert {:safe_email, "email not pwned"} = Pwned.check_email(@test_safe_email)
    end
  end
end


test/pwned/password_test.exs
defmodule Pwned.PasswordTest do
  use ExUnit.Case, async: true
  alias Pwned

  describe "verify check_password/1 returns an {:ok, count} tuple" do
    test "check_password returns the password pwned count" do
      assert {:ok, _} = Pwned.check_password("P@ssw0rd")
    end

    test "check_password pwned count is an integer" do
      {:ok, count} = Pwned.check_password("P@ssw0rd")

      assert is_integer(count)
    end
  end
end


lib/pwned.ex
defmodule Pwned do
  @moduledoc """
  Check if your password has been pwned.
  """
  alias Pwned.Utils.EmailFlattener
  alias Pwned.Utils.EmailReducer

  @doc """
  It uses [have i been pwned?](https://haveibeenpwned.com) to verify if a password has appeared in a data breach. In order to protect the value of the source password being searched the value is not sended through the network.

  ## Examples

      iex> Pwned.check_password("P@ssw0rd")
      {:ok, 47205}

      iex> Pwned.check_password("Z76okiy2X1m5PFud8iPUQGqusShCJhg")
      {:ok, false}

  """
  @spec check_password(String.t()) :: {:ok, integer} | {:ok, false} | :error
  def check_password(password) do
    with {head, rest} <- hash(password),
         {:ok, response} <- range_client().get(head),
         {:ok, range} <- parse_password_response(response),
         {:ok, answer} <- do_check(range, rest) do

      {:ok, answer}
    else
      :error -> :error
    end
  end

  @doc """
  This uses API v3 of the [have i been pwned?](https://haveibeenpwned.com) to check
  if an email has appeared in a data breach. It returns the total count of appearances.

  It also requires a purchased hibp-api-key, and implements a 12-factor methodology by
  accessing the hibp-api-key from the system's environment variables.

  ## Examples

    iex> Pwned.check_email("test123@example.com")
    {:pwned_email, 4893554722}

    iex Pwned.check_email("Z76okiy2X1m5PFud8iPUQGqusShCJhg@example.com")
    {:safe_email, "email not pwned"}

  """
  def check_email(email) do
    with head <- email,
         {:pwned_email, response} <- api_client().get(head),
         {:ok, response} <- parse_email_response(response),
         email_list <- EmailFlattener.flatten(response),
         pwned_count <- EmailReducer.reduce_email_list(email_list) do

      {:pwned_email, pwned_count}
    else
      {:safe_email, message} -> {:safe_email, message}

      {:error, message} -> {:error, message}

      _ -> :error
    end
  end

  defp hash(password) do
    :crypto.hash(:sha, password)
    |> Base.encode16()
    |> String.split_at(5)
  end

  defp parse_password_response(response) do
    parsed_response =
      response
      |> String.split("\r\n")
      |> Enum.map(fn line -> String.split(line, ":") end)

    {:ok, parsed_response}
  end

  defp parse_email_response(response) do
    parsed_response = Regex.scan(~r/"PwnCount":\d+/, response)
    {:ok, parsed_response}
  end

  defp do_check(range, rest) do
    case find_hash(range, rest) do
      {:ok, false} -> {:ok, false}
      {:ok, count} -> parse_count(count)
    end
  end

  defp find_hash(range, hash) do
    range
    |> Enum.find(fn item -> List.first(item) == hash end)
    |> handle_hash()
  end

  defp handle_hash(nil), do: {:ok, false}
  defp handle_hash([_hash, count]), do: {:ok, count}

  defp parse_count(count) do
    Integer.parse(count)
    |> handle_count()
  end

  defp handle_count(:error), do: :error
  defp handle_count({count, _rest}), do: {:ok, count}

  defp range_client, do: Application.get_env(:pwned_coretheory, :range_client, Pwned.Range.HTTPClient)
  defp api_client, do: Application.get_env(:pwned_coretheory, :api_client, Pwned.Utils.APIClient)
end


lib/utils/api_client.ex
defmodule Pwned.Utils.APIClient do
  @moduledoc """
  Checks if a given email has been pwned using
  the HaveIBeenPwned? API v3 [breachedaccounts service](https://haveibeenpwned.com/API/v3#BreachesForAccount).

  Requires a purchased [hibp-api_key](https://haveibeenpwned.com/API/Key).
  """

  @user_agent Application.compile_env(:pwned_coretheory, :user_agent, "Pwned by Core Theory Elixir Client")

  def get(head) do
    hibp_api_key = System.get_env("HIBP_API_KEY")

    case HTTPoison.get("https://haveibeenpwned.com/api/v3/breachedaccount/#{head}?truncateResponse=false", [{"hibp-api-key", hibp_api_key}, {"user-agent", @user_agent}]) do
      {:ok, %HTTPoison.Response{body: body, status_code: 200}} ->
        {:pwned_email, body}

      {:ok, %HTTPoison.Response{body: _, status_code: 404}} ->
        {:safe_email, "email not pwned"}

      {:ok, response} ->
        expected_api_response_errors(response)

      {:error, %HTTPoison.Error{id: nil, reason: :timeout}} ->
        {:error, "request_timeout"}

      _ ->
        :error
    end
  end

  def expected_api_response_errors(response) do
    case response do
      %HTTPoison.Response{body: _, status_code: 400} ->
        {:error, "bad request"}

      %HTTPoison.Response{body: _, status_code: 401} ->
        {:error, "unauthorized api key"}

      %HTTPoison.Response{body: _, status_code: 403} ->
        {:error, "forbidden user agent"}

      %HTTPoison.Response{body: _, status_code: 429} ->
        {:error, "too many requests"}

      %HTTPoison.Response{body: _, status_code: 503} ->
        {:error, "service unavailable"}
    end
  end
end


lib/utils/email_reducer.ex
defmodule Pwned.Utils.EmailReducer do
  @moduledoc """
  Functions for reducing the email list returned
  from the [breachedaccount](https://haveibeenpwned.com/API/v3#BreachesForAccount) API v3.
  """

  @doc """
  Reduces the list of flattened emails returned from the
  `EmailFlattener.flatten/1` call, and returns the
  integer sum.

  This sum indicates the number of times a given email has
  appeared in data breaches from the [breachedaccount](https://haveibeenpwned.com/API/v3#BreachesForAccount)
  service.

  ## Example
  iex> reduce_email_list(email_list)
        4893554722
  """
  def reduce_email_list(email_list) do
    email_list
    |> to_string()
    |> String.replace("\"PwnCount\"", "")
    |> String.split(":", trim: true)
    |> Stream.map(fn count -> String.to_integer(count) end)
    |> Enum.reduce(fn count, acc -> count + acc end)
  end
end


mix.exs
defmodule Pwned.MixProject do
  use Mix.Project

  @version "1.5.3"

  def project do
    [
      app: :pwned_coretheory,
      version: @version,
      elixir: "~> 1.11",
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      description: description(),
      package: package(),
      docs: docs(),
      source_url: "https://github.com/coretheory/pwned_coretheory"
    ]
  end

  def application do
    [
      applications: [:httpoison],
      extra_applications: [:logger],
    ]
  end

  defp deps do
    [
      {:httpoison, "~> 1.8"},
      {:ex_doc, "~> 0.23.0", only: :dev, runtime: false},
      {:dialyxir, "~> 1.0", only: [:dev, :test], runtime: false},
      {:credo, "~> 1.6", only: [:dev, :test], runtime: false},
      {:sobelow, "~> 0.11", only: :dev},
    ]
  end

  defp description() do
    """
    A simple application to check if an email or password has been pwned
    using the HaveIBeenPwned? API. It requires a purchased hibp-api-key
    in order to use the email checking functions.
    """
  end

  defp package() do
    [
      maintainers: ["Core Theory"],
      licenses: ["Apache 2.0"],
      links: %{"GitHub" => "https://github.com/coretheory/pwned_coretheory"},
    ]
  end

  defp docs() do
    [
      main: "readme",
      name: "Pwned by Core Theory",
      source_ref: "v#{@version}",
      canonical: "https://hexdocs.pm/pwned_coretheory/",
      source_url: "https://github.com/coretheory/pwned_coretheory",
      logo: "assets/static/images/CT_Logo_Color.png",
      extras: [
        "README.md",
        "CHANGELOG.md",
        "CONTRIBUTING.md",
        "FURTHER_READING.md",
        "LICENSE.md"
      ]
    ]
  end
end
